projects/utils-maps/src/lib/ol/utils-ol-layers.ts

```diff
@@ -0,0 +1,287 @@
+  const layer = getLayersByKey(map, key, value, lowerType, filtertypeKey);
+  if (Array.isArray(layer)) {
+    layer.forEach(l => map.removeLayer(l));
+  } else {
+    map.removeLayer(layer);
+  }
+}
+
+
+/**
+ * Removes all Layers or LayerGroups or for a filtertype.
+ *
+ * @param filtertypeKey [filtertypeKey='filtertype']
+ */
+export function removeAllLayers(map: Map, filtertype?: string, filtertypeKey = FILTER_TYPE_KEY) {
+  const layers = getLayers(map, filtertype, filtertypeKey);
+  if (filtertype) {
+    layers.forEach((lg) => {
+      if (lg.get(filtertypeKey) && lg.get(filtertypeKey).toLowerCase() === filtertype.toLowerCase()) {
+        map.removeLayer(lg);
+      }
+    });
+  } else {
+    map.getLayerGroup().getLayers().clear();
+  }
+}
+
+
+/**
+ * Removes a Layer/Group from a LayerGroup - by layer.get(key)
+ * @param key [key='id']
+ */
+export function removeLayerFromGroup(layer: BaseLayer, group: LayerGroup, key: string = 'id') {
+  const layers = getLayersFromGroup(group);
+  const filteredLayers = layers.filter(l => l.get(key) !== layer.get(key));
+  group.setLayers(new Collection(filteredLayers));
+}
+
+
+/**
+ * Get a flat array of layers from all layers and groups
+ */
+export function flattenLayers(layers: BaseLayer[]): Layer<any>[] {
+  let flattenedLayers: Layer<any>[] = [];
+  for (const layer of layers) {
+    if (layer instanceof LayerGroup) {
+      const subLayers = layer.getLayers().getArray();
+      const flattenedSubLayers = flattenLayers(subLayers);
+      flattenedLayers = Array.prototype.concat(flattenedLayers, flattenedSubLayers);
+    } else {
+      // Casting is terrible, but this cast is ok for once: since `layer` is no LayerGroup, it must be a Layer.
+      // See the `Subclasses` section here: https://openlayers.org/en/latest/apidoc/module-ol_layer_Base-BaseLayer.html
+      flattenedLayers.push(layer as Layer<any>);
+    }
+  }
+  return flattenedLayers;
+}
+
+
+/** --------------------------------------------------------------- */
+
+export function resolutionsFromExtent(extent, optMaxZoom: number, tileSize: number) {
+  const maxZoom = optMaxZoom;
+
+  const height = olGetHeight(extent);
+  const width = olGetWidth(extent);
+
+  const maxResolution = Math.max(width / tileSize, height / tileSize);
+
+  const length = maxZoom + 1;
+  const resolutions = new Array(length);
+  for (let z = 0; z < length; ++z) {
+    resolutions[z] = maxResolution / Math.pow(2, z);
+  }
+  return resolutions;
+}
+
+
+export function matrixIdsFromResolutions(resolutionLevels: number, matrixIdPrefix?: string) {
+  return Array.from(Array(resolutionLevels).keys()).map(l => {
+    if (matrixIdPrefix) {
+      return `${matrixIdPrefix}:${l}`;
+    } else {
+      return l;
+    }
+  });
+}
+
+
+
+export function getTileGrid<T>(type: 'wmts' | 'default' = 'default', resolutionLevels?: number, tileSize?: number, matrixIdPrefix?: string, resolutions?: Array<string | number>, matrixIds?: Array<string | number>): T {
```

Comment:
> `getTileGrid<T>(type: 'wmts' | 'default' = 'default'`
This seems a little odd to me. `T` must be either `TileGrid` or `WMTSTileGrid`. Some parameters (`matrixIdPrefix` and `matrixIds` only apply to the latter. I feel like the consumer of this library should not have to know those internal details. 

@boeckMt : What do you think: should we split this method into two? 
 - `getTileGrid(tileSize?: number, resolutions?: number[], resolutionLevels?: number)`
 - `getWMTSTileGrid(tileSize?: number, resolutions?: number[], resolutionLevels?: number, matrixIds?: string[], matrixIdPrefix?: string)`

Reply:
Coming to think of it: `resolutions` and `resolutionLevels` are mutually exclusive. So are `matrixIds` and `matrixIdPrefix`.  I feel like those should be separate functions, too. So how about those new methods: 

 - `getTileGrid(resolutions: number[], tileSize?: number)`
 - `getTileGridAuto(resolutionLevels: number, tileSize?: number)`
 - `getWMTSTileGrid(resolutions: number[], matrixIds: string[], tileSize?: number )`
 - `getWMTSTileGridAuto(resolutionLevels: number, matrixIdPrefix: string, tileSize?: number )`